### ISSUE TITLE
perf: Performance bottlenecks in FuturesOneStepEnv and MarketDataObservationSampler

### ISSUE THREAD

**Issue by BY571** [TorchTrade/torchtrade#36]
## Summary

Deep investigation into performance bottlenecks affecting online training speed for `FuturesOneStepEnv`. Found multiple critical issues that could yield **5-10x speedup** if addressed.

---

### 1. Redundant `get_base_features()` calls in `_rollout()` loop

**File:** `torchtrade/envs/offline/futuresonestepenv.py:608-611`

```python
while not self.truncated:
    obs_dict, self.current_timestamp, self.truncated = self.sampler.get_sequential_observation()
    ohlcv = self.sampler.get_base_features(self.current_timestamp)  # Called EVERY iteration
    self._cached_base_features = ohlcv
```

**Problem:** For each bar in the rollout (potentially 100+ steps), `get_base_features()` performs:
- Timestamp conversion to int64 nanoseconds
- `torch.searchsorted()` on the execute_base_idx tensor
- Tensor slicing and `.tolist()` conversion

**Impact:** 10,000 episodes × 50 steps average = **500,000+ redundant searchsorted operations** per training run.

**Fix:** The OHLCV data is already available from `get_sequential_observation()`. Cache it or return it directly.

---

### 2. Multiple `torch.searchsorted()` per observation (4x for 4 timeframes)

**File:** `torchtrade/envs/offline/sampler.py:140-175`

```python
for tf, ws in zip(self.time_frames, self.window_sizes):
    # Each iteration does its own binary search for the SAME timestamp
    pos_t = torch.searchsorted(idx_tensor, ts_t, right=True)
```

**Problem:** All timeframes share the same execution timestamp. Binary search position could be computed ONCE then mapped to each timeframe via a precomputed offset table.

**Impact:** 4 TF × 100K steps = 400K binary searches that could be reduced to 100K.

**Fix:** Create a master index or precompute position mappings between timeframes.

---

### 3. Tensor creation in `compute_return()` - inside rollout loop

**File:** `torchtrade/envs/offline/futuresonestepenv.py:512-520`

```python
def compute_return(self, close_price: float) -> torch.Tensor:
    current_value = self._get_portfolio_value(close_price)
    current_return = torch.log(
        torch.tensor(current_value, dtype=torch.float) /  # NEW tensor
        torch.tensor(self.previous_portfolio_value, dtype=torch.float)  # NEW tensor
    )
    return current_return
```

**Problem:** Called once per rollout step, creating TWO new tensors just to compute a scalar log return. For a 50-step rollout: 100 tensor creations.

**Fix:** Use `math.log(current_value / self.previous_portfolio_value)` and convert to tensor only at the end.

---

### 4. Double GPU transfer + `.item()` blocking in training loop

**File:** `examples/online/grpo_futures/train.py:156-172`

```python
data = next(collector_iter).to(device)  # First transfer (no non_blocking)
metrics_to_log["train/action_std"] = data["action"].float().std().item()  # .item() BLOCKS GPU

def update(batch):
    batch = batch.to(device, non_blocking=True)  # SECOND transfer - data already on device!
```

**Problem:**
- Line 157 transfers data to device **without** `non_blocking=True`
- Line 161: `.item()` blocks the entire GPU waiting for action_std computation
- Line 129: Redundant second transfer (data already on device)

**Fix:** Add `non_blocking=True` to first transfer, remove redundant second transfer, defer `.item()` calls.

---

### 5. Window slicing creates tensor copies

**File:** `torchtrade/envs/offline/sampler.py:168`

```python
window = arr[start: idx_pos + 1]  # Creates new tensor with copied data
```

**Impact:** 4 TF × 52 bars × 20 features × 4 bytes = **16.6 KB per step**. For 100K steps = 1.66 GB of redundant tensor copies.

**Fix:** Pre-allocate observation buffers and copy into them, or use tensor views where possible.

---

### 6. Sequential `_rollout()` loop - no vectorization

**File:** `torchtrade/envs/offline/futuresonestepenv.py:592-664`

The rollout loop is completely sequential:
- Cannot be batched across environments
- Variable number of iterations (1-100s of steps)
- Each iteration has multiple dict lookups and conditional checks

**Fix:** Consider vectorizing the SL/TP price checks, or pre-allocating rollout buffers.

---

### 7. Duplicate portfolio value calculations

**File:** `torchtrade/envs/offline/futuresonestepenv.py:477-494`

`_get_observation()` calculates `position_value`, `unrealized_pnl`, etc., then `_get_portfolio_value()` recalculates `unrealized_pnl` again.

**Fix:** Return portfolio value from `_get_observation()` or cache the calculation.

---

### 8. Unnecessary transforms on one-step environment

**File:** `examples/online/grpo_futures/utils.py:199-210`

```python
Compose(
    InitTracker(),      # Overhead for one-step envs
    DoubleToFloat(),    # Data already float32
    RewardSum(),        # Unnecessary - one-step returns done=True every step
    StepCounter(max_steps=max_steps),
)
```

**Fix:** Remove `RewardSum()` and `DoubleToFloat()` for one-step environments.

---

### 9. Account state tensor allocation every step

**File:** `torchtrade/envs/offline/futuresonestepenv.py:369-381`

Creates a new 10-element tensor on every step. Could pre-allocate and update in-place.

---

### 10. No pinned memory for data movement

**File:** `examples/online/grpo_futures/train.py:157`

Data transfers don't use pinned memory, resulting in ~30-50% slower PCIe transfers.

---

### 11. Cascading DataFrame copies during sampler init

**File:** `torchtrade/envs/offline/sampler.py:64-71`

```python
resampled = feature_processing_fn(resampled)  # Copy 1
resampled = resampled[cols_to_keep]  # Copy 2
resampled = resampled.reset_index().set_index("timestamp")  # Copy 3 & 4
resampled = resampled.drop(columns=["index"])  # Copy 5
```

**Impact:** 5 DataFrame copies per timeframe during initialization.

---

## Priority Recommendations

| Priority | Issue | Estimated Speedup | Effort |
|----------|-------|-------------------|--------|
| 1 | Remove redundant `get_base_features()` in rollout | 5-10x for rollout | Low |
| 2 | Fix double GPU transfer + .item() blocking | 2-3x for training loop | Low |
| 3 | Replace tensor creation in `compute_return()` | 2x for reward calc | Low |
| 4 | Single searchsorted for all timeframes | 2-4x for observations | Medium |
| 5 | Pre-allocate observation buffers | 1.5-2x for observations | Medium |
| 6 | Remove unnecessary transforms | 1.2x per step | Low |

**Total estimated speedup: 5-10x** if critical issues are addressed.
